### ISSUE TITLE
Tracing API

### ISSUE THREAD

**Issue by markrwilliams** [glyph/automat#36]
We should work out an API that allows you to trace the execution of an Automat machine.  You should able to register with a machine an object that receives a sequence of `(input, state, outputs)` tuples and then unregister it.  It should help with [other issues](https://github.com/glyph/automat/issues/8).

**Comment by warner** [comment]
I've got a need for this, so I tried to dive in. I only got as far as writing a test case, but it brought up a question of what exactly the registered function should get.

For debugging my application, I want to see a series of lines like:

* state1 -> INPUT1 -> state2
* state2 -> INPUT2 -> state3

I'm also interested in the outputs that get called, and how they're interleaved with other events in my program. But not always: the coarsest level of logging would just be the main state transitions.

So maybe the tracing function should take `(old_state, input, new_state, output_or_None)`. It would be called at least once (with `output == None`) for each state transition (when the input first arrives, before any outputs have been executed). Then it would be called again for each output (same first three args, but a new `output=` for each that gets run).

That way, for coarse logging, I could only print a line if `output is None`, and for detailed logging I'd print a line for every call to the tracing function.

Does that line up with what you were thinking? I don't know how the `(input, state, outputs)` signature that you mentioned would work.

The test case would look something like this:

```python
from unittest import TestCase
from .._methodical import MethodicalMachine

class SampleObject(object):
    mm = MethodicalMachine()

    @mm.state(initial=True)
    def begin(self):
        "initial state"
    @mm.state()
    def middle(self):
        "middle state"
    @mm.state()
    def end(self):
        "end state"

    @mm.input()
    def go1(self):
        "sample input"
    @mm.input()
    def go2(self):
        "sample input"
    @mm.input()
    def back(self):
        "sample input"

    @mm.output()
    def out(self):
        "sample output"

    begin.upon(go1, middle, [out])
    middle.upon(go2, end, [out])
    end.upon(back, middle, [])
    end.upon(back, begin, [])

class TraceTests(TestCase):
    def test_trace(self):
        traces = []
        def tracer(old_state, input, new_state, output):
            traces.append((old_state, input, new_state, output))
        s = SampleObject()
        s.mm.setTrace(tracer)

        s.go1()
        self.assertEqual(traces, [(s.begin, s.go1, s.middle, None),
                                  (s.begin, s.go1, s.middle, s.out),
                                  ])
        s.go2()
        self.assertEqual(traces, [(s.begin, s.go1, s.middle, None),
                                  (s.begin, s.go1, s.middle, s.out),
                                  (s.middle, s.go2, s.end, None),
                                  (s.middle, s.go2, s.end, s.out),
                                  ])
        s.mm.setTrace(None)
        s.back()
        self.assertEqual(traces, [(s.begin, s.go1, s.middle, None),
                                  (s.begin, s.go1, s.middle, s.out),
                                  (s.middle, s.go2, s.end, None),
                                  (s.middle, s.go2, s.end, s.out),
                                  ])
        s.go2()
        self.assertEqual(traces, [(s.begin, s.go1, s.middle, None),
                                  (s.begin, s.go1, s.middle, s.out),
                                  (s.middle, s.go2, s.end, None),
                                  (s.middle, s.go2, s.end, s.out),
                                  ])
```

I also need to find a way to transform the `MethodicalInput`/etc objects into a nice printable representation, one which doesn't include all the python object IDs and `attr.ib` details. Ideally just a single word.

And I haven't figured out how to add a method to `MethodicalMachine` when it tries to hide itself as an "implementation detail": perhaps there's some decorator-ish way to add a method to the host object which can be used to activate tracing. Maybe something like:

```python
class SampleObject(object)
    mm = MethodicalMachine()

    @mm.setTrace()
    def call_this_to_setTrace(self, tracer):
        # body is never actually called
        pass

so = SampleObject()
so.call_this_to_setTrace(my_tracer_function)
```

**Comment by warner** [comment]
Take a look at that commit, maybe tell me if I'm headed in the right direction.

* It uses the `@mm.setTrace()` decorator as above, which is a bit awkward. I think I now understand why MethodicalMachine is trying so heroically to hide itself, but it does make it hard to just talk to the state machine instance when you want to. I think it should be possible to do `call_this_to_setTrace = mm.setTrace()` instead of the decorator, but I'll have to understand how `__get__` works properly to do it (I just copied/pasted the MethodicalInput decorator to achieve `@mm.setTrace()`), and then I wonder if seeing `mm.something` *without* the `@` is going to stand out and cause confusion.
* It passes strings into the tracing function, rather than decorated methods. I think this is more useful in practice: you can print them without diving into the internals to extract a method name. And the method names have to be unique anyways. 
* It does not yet capture the outputs (so the tracing function in my branch takes 3-tuples, not 4-tuples). I'm still trying to find a clean way to get the tracing information (the function, and the names of the states and input) into the place where the outputs are called, so the tracer can be invoked just before each output is invoked. Currently that place does not need to know about states and inputs, so it doesn't come near them.

I'm thinking about adding a spare context arg/kwarg to the setTrace call, sort of like how `Defer.addDeferred(cb, "foo", bar="baz")` results in `cb(result, "foo", bar="baz")`. For tracing, this is where you'd record which machine you're logging (and which instances of it, if you've got multiple FooMachines in the same process), so your log doesn't become an undifferentiated stream of "state=begin -> input=start -> state=started" because all your various machines have such stereotypical names.

But it might be easier/better to require the caller to capture that data in a lambda or a closure, and not pass additional args into the registration function. The tracer isn't going to be set all that often, they can afford declaring an extra function or two. (For Deferreds I can see a different argument, I call .addCallback with an extra context argument so frequently that it'd be annoying to wrap each one in a closure to get the context). Thoughts?

**Comment by glyph** [comment]
This generally looks like the right idea; I assume you can work through all the metaprogramming quirks yourself, but _design_-wise I think this is roughly going down the right track.  I think you mean `Deferred.addCallback`, though? :)

**Comment by warner** [comment]
Closed in #56 , with a private API that you use with `do_set_trace = mm._setTrace`, which lets you register a function that accepts strings (old_state, input, new_state), and can optionally return a callable that lets you follow per-output invocations too. Full docs in "debugging.md".
